smolyay/basis.py

```diff
@@ -310,6 +337,38 @@ def __call__(self,x):
                 answer += math.comb(self._n+1,2*k+1)*((x**2 - 1)**k)*(x**(self._n-2*k))
             return answer
 
+    def derivative(self, x):
+        r"""Evaluate the derivative of Chebyshev Second Kind.
+
+        The first derivative of Chebyshev polynomials of second kind is
+        evaluated using the connection between Chebyshev polynomial of
+        first kind and second kind.
+
+        Parameters
+        ----------
+        x: float
+            input in [-1, 1] domain.
+
+        Returns
+        -------
+        float
+            Value of the derivative of Chebyshev polynomials of first kind.
+
+        Raises
+        ------
+        ValueError
+            if input is outside the domain [-1,1].
+        """
+        if x > 1 or x < -1:
+            raise ValueError("Input is outside the domain [-1,1]")
+        if x == 1:
+            return self._n*(self._n+1)*(self._n+2)/3
+        elif x == -1:
+            return ((-1)**(self._n+1))*self._n*(self._n+1)*(self._n+2)/3
+        else:
+            return ((self._n+1)*ChebyshevFirstKind(self._n+1)(x) -
+                    x*ChebyshevSecondKind(self._n)(x))/(x**2-1)
```

Comment:
Per our discussion on Slack, you should not make a new polynomial each time. Instead, you can make T_{n+1}, then use the call method of the self for the second kind::
```suggestion
            if self._derivative_polynomial is None:
                self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
            return ((self._n+1)*self._derivative_polynomial(x) -
                    x*self(x))/(x**2-1)
```
You should set `self._derivative_polynomial = None` in the constructor.

You can do a similar `if not None` syntax in the `ChebyshevFirstKind`, if you'd like to delay making the polynomial until the derivative is evaluated. It would make sense to do both in the same way.